addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/internal/discovery/NeeoRoomDiscoveryService.java

```diff
@@ -0,0 +1,237 @@
+     */
+    @Reference(cardinality = ReferenceCardinality.MANDATORY, service = ThingRegistry.class, name = "ThingRegistry", policy = ReferencePolicy.DYNAMIC, unbind = "unsetThingRegistry")
+    protected void setThingRegistry(ThingRegistry thingRegistry) {
+        Objects.requireNonNull(thingRegistry, "thingRegistry cannot be null");
+
+        this.thingRegistry = thingRegistry;
+    }
+
+    /**
+     * Unsets thing registry.
+     *
+     * @param thingRegistry the thing registry (ignored)
+     */
+    protected void unsetThingRegistry(ThingRegistry thingRegistry) {
+        this.thingRegistry = null;
+    }
+
+    /**
+     * Set's the NEEO type generator
+     *
+     * @param typeGenerator a possibly null {@link NeeoTypeGenerator}
+     */
+    @Reference(cardinality = ReferenceCardinality.MANDATORY, service = NeeoTypeGenerator.class, name = "NeeoTypeGenerator", policy = ReferencePolicy.DYNAMIC, unbind = "unsetNeeoTypeGenerator")
+    protected void setNeeoTypeGenerator(NeeoTypeGenerator typeGenerator) {
+        Objects.requireNonNull(typeGenerator, "typeGenerator cannot be null");
+
+        this.typeGenerator = typeGenerator;
+    }
+
+    /**
+     * Unsets neeo type generator
+     *
+     * @param typeGenerator the neeo type provider (ignored)
+     */
+    protected void unsetNeeoTypeGenerator(NeeoTypeGenerator typeGenerator) {
+        this.typeGenerator = null;
+    }
+
+    @Override
+    public void deactivate() {
+        logger.debug("Deactivating any active scans");
+        NeeoUtil.cancel(scan.getAndSet(null));
+    }
+
+    @Override
+    protected void startScan() {
+        logger.debug("Starting NEEO Room scan");
+        NeeoUtil.cancel(scan.getAndSet(scheduler.submit(() -> {
+            try {
+                scanForRooms();
+            } catch (InterruptedException e) {
+                logger.debug("NEEO Room scan interrupted: {}", e.getMessage(), e);
+            }
+        })));
+    }
+
+    /**
+     * Starts the scan for rooms by looking through the {@link ThingRegistry} for any thing that has a
+     * {@link NeeoBrainHandler} as it's handler. We then create {@link DiscoveryResult} for each room in the brain.
+     *
+     * @throws InterruptedException if our scan is interrupted
+     */
+    private void scanForRooms() throws InterruptedException {
+        NeeoUtil.checkInterrupt();
+
+        final ThingRegistry localThingRegistry = this.thingRegistry;
+        if (localThingRegistry == null) {
+            logger.debug("ThingRegistry is null - scan aborted");
+            return;
+        }
+
+        final NeeoTypeGenerator localTypeGenerator = this.typeGenerator;
+        if (localTypeGenerator == null) {
+            logger.debug("TypeGenerator is null - scan aborted");
+            return;
+        }
+
+        logger.debug("Scanning all things for NeeoBrainHandlers");
+        for (final Thing thing : localThingRegistry.getAll()) {
```

Comment:
Since you do not have the thingregistry anymore you can create a discovery service from the handler factory and pass the brain bridge to it so it can access its api and query the hardware directly.